plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergV2.java

```diff
@@ -339,6 +346,87 @@ public void testOptimizingPartitionsOfV2TableWithGlobalEqualityDeleteFile()
                                 .toArray(String[]::new));
     }
 
+    @Test(timeOut = 90_000)
+    public void testOptimizeDuringWriteOperations()
+            throws Exception
+    {
+        int invocationCount = 3;
+        boolean optimizeSucceeded = false;
+        for (int attempt = 0; attempt < invocationCount; attempt++) {
+            optimizeSucceeded = runOptimizeDuringWriteOperations();
+            if (optimizeSucceeded) {
+                break;
+            }
+        }
+        assertTrue(optimizeSucceeded, "Expected optimize to succeed in at least one test attempt");
+    }
+
+    private boolean runOptimizeDuringWriteOperations()
+            throws Exception
+    {
+        int threads = 5;
+        int deletionThreads = threads - 1;
+        int rows = 100;
+        CyclicBarrier barrier = new CyclicBarrier(threads);
+        ExecutorService executor = newFixedThreadPool(threads);
+        try (TestTable table = new TestTable(
+                getQueryRunner()::execute,
+                "test_optimize_during_write_operations",
+                "(int_col INT)")) {
+            String tableName = table.getName();
+            for (int i = 0; i < rows; i++) {
+                assertUpdate(format("INSERT INTO %s VALUES %s", tableName, i), 1);
```

Comment:
inserting them with one insert, would be much faster (eg use `sequence`)

if you need 100 data files, add a code comment

Reply:
100 is definitely overkill, but I did need individual files to hit the situation that was causing the problem because it has to do with the optimization we have to use metadata deletes when a file is fully removed.